config/webpack.config.dev.js

```diff
@@ -134,10 +135,41 @@ module.exports = {
       // We are waiting for https://github.com/facebookincubator/create-react-app/issues/2176.
       // { parser: { requireEnsure: false } },
 
+      // First, run the linter.
+      // It's important to do this before Babel processes the JS.
       {
-        test: /\.(js|jsx|mjs)$/,
-        loader: require.resolve('source-map-loader'),
+        test: /\.(js|jsx|mjs|ts|tsx)$/,
         enforce: 'pre',
+        use: [
+          require.resolve('source-map-loader'),
+          {
+            options: {
+              formatter: eslintFormatter,
+              eslintPath: require.resolve('eslint'),
+              baseConfig: {
+                extends: [require.resolve('eslint-config-react-app')],
+                parser: 'typescript-eslint-parser',
+                overrides: [
+                  {
+                    files: ['*.ts', '*.tsx'],
+                    rules: {
+                      'no-array-constructor': 'off',
+                      'no-multi-str': 'off',
+                      'no-restricted-globals': 'off',
+                      'no-undef': 'off',
+                      'no-unused-vars': 'off',
+                    },
+                  },
+                ],
+              },
+              // @remove-on-eject-begin
+              ignore: false,
+              useEslintrc: false,
```

Comment:
CRA also disables ESLint customization. My intention was to follow their decisions as closely as possible to simplify maintenance and take advantage of CRA's feature set, familiarity, and ease of use.

> Note that even if you edit your .eslintrc file further, these changes will only affect the editor integration. They won’t affect the terminal and in-browser lint output.

I think what you're referring to is editor configuration. That is still possible both in CRA and this PR, but it won't affect the Webpack loader. The configs can be customized after ejecting, but CRA intentionally discourages this to simplify its tooling to common use cases.

> why configure it in the loader (hardcoded and twice - dev/prod), instead of using the standard .eslintrc?

I did this because it was easier, but I agree it's not ideal. I think we should set up a custom ESLint config similar to eslint-config-react-app with the necessary TS tweaks, extending the original. Unfortunately I can't do that in the current repository format because it now only has one package, so we'll have to decide on the best way to approach it.

Reply:
I know it seems restrictive, but it's not too difficult to add `eslint src` to your npm scripts with an extended config. That's what I personally do with CRA. https://github.com/nickmccurdy/password-entropy/blob/master/package.json#L35